lib/inherited_resources/url_helpers.rb

```diff
@@ -184,56 +174,71 @@ def generate_url_and_path_helpers(prefix, name, resource_segments, resource_ivar
       singleton   = self.resources_configuration[:self][:singleton]
       polymorphic = self.parents_symbols.include?(:polymorphic)
 
-      # If it's not a singleton, ivars are not empty, not a collection or
-      # not a "new" named route, we can pass a resource as argument.
-      #
-      unless (singleton && name != :parent) || ivars.empty? || name == :collection || prefix == :new
-        ivars.push "(given_args.first || #{ivars.pop})"
-      end
-
       # In collection in polymorphic cases, allow an argument to be given as a
       # replacemente for the parent.
       #
-      if name == :collection && polymorphic
-        index = ivars.index(:parent)
-        ivars.insert index, "(given_args.first || parent)"
-        ivars.delete(:parent)
-      end
+      parent_index = ivars.index(:parent) if polymorphic
 
-      # When polymorphic is true, the segments must be replace by :polymorphic
-      # and ivars should be gathered into an array, which is compacted when
-      # optional.
-      #
-      if polymorphic
-        segments = :polymorphic
-        ivars    = "[#{ivars.join(', ')}]"
-        ivars   << '.compact' if self.resources_configuration[:polymorphic][:optional]
+      segments = if polymorphic
+        :polymorphic
+      elsif resource_segments.empty?
+        'root'
       else
-        segments = resource_segments.empty? ? 'root' : resource_segments.join('_')
-        ivars    = ivars.join(', ')
+        resource_segments.join('_')
       end
 
-      prefix = prefix ? "#{prefix}_" : ''
-      ivars << (ivars.empty? ? 'given_options' : ', given_options')
+      ivars = ivars.present? ? Array(ivars) : []
 
-      given_args_transform = 'given_args = given_args.collect { |arg| arg.respond_to?(:permitted?) ? arg.to_h : arg }'
+      define_params_helper(prefix, name, singleton, polymorphic, parent_index, ivars)
+      [:path, :url].each { |suffix| define_helper_method(prefix, name, suffix, segments) }
+    end
 
-      class_eval <<-URL_HELPERS, __FILE__, __LINE__
-        protected
-          undef :#{prefix}#{name}_path if method_defined? :#{prefix}#{name}_path
-          def #{prefix}#{name}_path(*given_args)
-            #{given_args_transform}
-            given_options = given_args.extract_options!
-            #{prefix}#{segments}_path(#{ivars})
-          end
+    def define_params_helper(prefix, name, singleton, polymorphic, parent_index, ivars)
+      params_method_name = ['', prefix, name, :params].compact.join(?_)
```

Comment:
We've introduced this `?_` new style for single character string literals. I didn't know about that style and I don't think many people do, and the change is not applied consistently anways so... I would keep the regular style.

Reply:
Wow. I didn't know you could do that! Very cool:
```
>> ['', 'prefix', 'name', :params].compact.join(?_)
=> "_prefix_name_params"
>> ?_
=> "_"
```

Agreed though about keeping the regular style.